src/02-assistant/index.ts

```diff
@@ -0,0 +1,261 @@
+   */
+
+  let assistant
+
+  if (process.env.ASSISTANT_ID) {
+    assistant = await openai.beta.assistants.retrieve(
+      process.env.ASSISTANT_ID as string
+    )
+  } else {
+    assistant = await openai.beta.assistants.create({
+      name: "Artsy Advisor",
+      instructions:
+        "You are an art advisor. Your job is to listen to your client and provide helpful recommendations of artworks and artists that they may like. You consider price range, medium, rarity, and other key attributes a client may want to consider when purchasing art. You ask clarifying questions where necessary to build an accurate profile on your client and provide more accurate recommendations.",
+      model: "gpt-3.5-turbo",
+      tools,
+    })
+  }
+
+  /*
+   * Create a new thread and send it a message with the user's input.
+   */
+
+  const thread = await openai.beta.threads.create()
+
+  await openai.beta.threads.messages.create(thread.id, {
+    role: "user",
+    content: input,
+  })
+
+  /*
+   * Run the thread using the desired assistant and poll until it reaches a completed state.
+   * https://platform.openai.com/docs/assistants/how-it-works/runs-and-run-steps
+   */
+
+  let run = await openai.beta.threads.runs.createAndPoll(thread.id, {
+    assistant_id: assistant.id,
+    instructions:
+      "Respond like you work at artsy.net. Always provide a list of artists and include the link to their profile. Always check artsy before making a recommendation.",
+  })
+
+  /*
+   * If the assistant wants to use a tool to help its response, then call the function (locally defined, further down) to get the result from Artsy's API.
+   */
+
+  if (
+    run.status === "requires_action" &&
+    run.required_action?.type === "submit_tool_outputs"
+  ) {
+    const name =
+      run.required_action.submit_tool_outputs.tool_calls?.[0].function.name
+    const args = JSON.parse(
+      run.required_action.submit_tool_outputs.tool_calls?.[0].function
+        .arguments || "null"
+    )
+
+    console.log(`Calling function: ${name} with args: ${JSON.stringify(args)}`)
+
+    let artists
+
+    if (name === "get_artists") {
+      artists = await get_artists(args)
+    }
+
+    if (name === "get_curated_artists") {
+      artists = await get_curated_artists(args)
+    }
+
+    if (artists) {
+      await openai.beta.threads.runs.submitToolOutputs(thread.id, run.id, {
+        tool_outputs: [
+          {
+            tool_call_id:
+              run.required_action.submit_tool_outputs.tool_calls?.[0].id,
+            output: JSON.stringify(artists, null, 2),
+          },
+        ],
+      })
+    }
+  }
+
+  /*
+   * Poll the run until it reaches a completed state and print the messages.
+   */
+
+  run = await openai.beta.threads.runs.poll(thread.id, run.id)
+
+  if (run.status === "completed") {
+    const messages = await openai.beta.threads.messages.list(run.thread_id)
+    for (const message of messages.data.reverse()) {
```

Comment:
**question (non-blocking):** Why `reverse` the messages here?

Reply:
This block came from their [docs](https://platform.openai.com/docs/assistants/overview/step-4-create-a-run), so not 100% but would guess is that by default the messages come in revers chronological order, which would be confusing when printed out.